attestation-agent/attestation-agent/src/lib.rs

```diff
@@ -0,0 +1,142 @@
+    async fn get_evidence(&mut self, runtime_data: &[u8]) -> Result<Vec<u8>>;
+
+    /// Extend runtime measurement register
+    async fn extend_runtime_measurement(
+        &mut self,
+        events: Vec<Vec<u8>>,
+        register_index: Option<u64>,
+    ) -> Result<()>;
+
+    /// Check the initdata binding
+    async fn check_init_data(&mut self, init_data: &[u8]) -> Result<()>;
+}
+
+/// Attestation agent to provide attestation service.
+pub struct AttestationAgent {
+    config: Option<Config>,
+}
+
+impl Default for AttestationAgent {
+    fn default() -> Self {
+        let config = Config::try_from(config::DEFAULT_AA_CONFIG_PATH).ok();
+        AttestationAgent { config }
+    }
+}
+
+impl AttestationAgent {
+    /// Create a new instance of [AttestationAgent].
+    pub fn new(config_path: &str) -> Self {
+        let config = Config::try_from(config_path).ok();
+
+        AttestationAgent { config }
+    }
+}
+
+#[async_trait]
+impl AttestationAPIs for AttestationAgent {
+    async fn get_token(&mut self, token_type: &str) -> Result<Vec<u8>> {
+        let _uri = match self.config.as_ref() {
+            Some(c) => c.as_uri.clone(),
+            None => {
+                let params = aa_kbc_params::get_params()
+                    .await
+                    .map_err(|_| anyhow!("Get AS URI failed"))?;
+                params.uri().to_string()
+            }
+        };
+
+        match TokenType::from_str(token_type).map_err(|e| anyhow!("Unsupported token type: {e}"))? {
+            #[cfg(feature = "kbs")]
+            token::TokenType::Kbs => token::kbs::KbsTokenGetter::default().get_token(_uri).await,
+            #[cfg(feature = "coco_as")]
+            token::TokenType::CoCoAS => {
+                token::coco_as::CoCoASTokenGetter::default()
+                    .get_token(_uri)
+                    .await
+            }
+        }
+    }
+
+    /// Get TEE hardware signed evidence that includes the runtime data.
+    async fn get_evidence(&mut self, runtime_data: &[u8]) -> Result<Vec<u8>> {
+        let tee_type = detect_tee_type();
+        let attester = TryInto::<BoxedAttester>::try_into(tee_type)?;
+        let evidence = attester.get_evidence(runtime_data.to_vec()).await?;
+        Ok(evidence.into_bytes())
+    }
+
+    /// Extend runtime measurement register
+    async fn extend_runtime_measurement(
+        &mut self,
+        events: Vec<Vec<u8>>,
+        register_index: Option<u64>,
+    ) -> Result<()> {
+        let tee_type = detect_tee_type();
+        let attester = TryInto::<BoxedAttester>::try_into(tee_type)?;
+        attester
+            .extend_runtime_measurement(events, register_index)
+            .await?;
+        Ok(())
+    }
+
+    /// Check the initdata binding. If current platform does not support initdata
+    /// injection, return a success and raise a warning log.
+    async fn check_init_data(&mut self, init_data: &[u8]) -> Result<()> {
```

Comment:
If we have 3 conceivable code paths, can we encode them into an enum?

Result<Ok<()> | Unsupported, anyhow::Error>

or

 Ok | Unsupported | anyhow::Error

Reply:
Nice idea. Let me add that.

Reply:
I added in `Result<Ok|Unsupported, Error>` way